src/test/scala/chiselTests/InstanceSpec.scala

```diff
@@ -0,0 +1,444 @@
+  }
+  "Template/Instance" should "enable instantiating nestingly, with modules" in {
+    import InstanceSpec.Examples._
+    import InstanceSpec.Annotations._
+    class AddOneTester extends BasicTester {
+      val template = Template(new AddTwo)
+      val i0 = Instance(template)
+      val i1 = Instance(template)
+      i0.in := 42.U
+      i1.in := i0.out
+      i0.i0.innerWire.map(x => mark(x, "Adam Was Here"))
+      i0.i1.innerWire.map(x => mark(x, "Megan Was Here"))
+      chisel3.assert(i1.out === 46.U)
+      stop()
+    }
+
+    val (output, annotations) = (new ChiselStage).emitChirrtlWithAnnotations(gen = new AddOneTester, args = Array("--full-stacktrace"))
+    annotations.toSeq should contain (MarkAnnotation(Target.deserialize("~AddOneTester|AddOneTester/i0:AddTwo/i0:AddOne>innerWire").asInstanceOf[ReferenceTarget], "Adam Was Here"))
+    annotations.toSeq should contain (MarkAnnotation(Target.deserialize("~AddOneTester|AddOneTester/i0:AddTwo/i1:AddOne_2>innerWire").asInstanceOf[ReferenceTarget], "Megan Was Here"))
+    assertTesterPasses(new AddOneTester)
+  }
+  "Template/Instance" should "implicitly convert modules to instances" in {
+    import InstanceSpec.Examples._
+    import InstanceSpec.Annotations._
+    def wireUp(i0: Instance[AddOne], i1: Instance[AddOne]): Unit = {
+      i1.in := i0.out
+    }
+    class AddOneTester extends BasicTester {
+      val template = Template(new AddOne(false))
+      val i0 = Instance(template)
+      val i1 = Module(new AddOne(true))
+      i0.in := 42.U
+      wireUp(i0, i1)
+      i1.in := i0.out
+      chisel3.assert(i1.out === 44.U)
+      stop()
+    }
+
+    val (output, annotations) = (new ChiselStage).emitChirrtlWithAnnotations(gen = new AddOneTester, args = Array("--full-stacktrace"))
+    assertTesterPasses(new AddOneTester)
+  }
+  "Template/Instance" should "work with top level declarations" in {
+    import InstanceSpec.Examples._
+    class AddZeroTester extends BasicTester {
+      val template = Template(new TopLevelDeclaration())
+      val i0 = Instance(template)
+      val template2 = Template(new TopLevelDeclarationWithCompanionObject())
+      val i1 = Instance(template2)
+      i0.in := 42.U
+      i1.in := i0.out
+      chisel3.assert(i1.out === 42.U)
+      stop()
+    }
+
+    val (output, annotations) = (new ChiselStage).emitChirrtlWithAnnotations(gen = new AddZeroTester, args = Array("--full-stacktrace"))
+    assertTesterPasses(new AddZeroTester)
+  }
+  "Template/Instance" should "work with public members of super classes" in {
+    @instance
+    class AddZero extends MultiIOModule {
+      @public val in  = IO(Input(UInt(32.W)))
+      @public val out = IO(Output(UInt(32.W)))
+      @public val clock = clock
+      out := in
+    }
+
+    @instance
+    class AddZeroWithCompanionObject extends MultiIOModule {
+      @public val in  = IO(Input(UInt(32.W)))
+      @public val out = IO(Output(UInt(32.W)))
+      @public val clock = clock
```

Comment:
I don't necessarily like the `@public` annotation. Wouldn't normally only the I/Os be public? Is there a way to automatically expose the I/Os without requiring a new annotation?

Reply:
i like the explicit annotation, but I don't think `@public` is the right word (it's too generic). `instancePublic` could be one suggestion, or `annotatable`

Reply:
So you are saying by default IOs should be public? That may require some cleverness to get working by default, but I can  work on it.

Reply:
should we distinguish between I/Os which might be optimized away and those that might not? Is there any implication if you mark an I/O `@to-be-renamed-public` on whether it can be optimized away?

Reply:
Optimizing away I/Os is a firrtl issue, not really a Chisel issue imho. This means that the firrtl compiler will be able to figure out which I/Os it can get rid of based on how the module is used, Chisel users should not have to worry about that.

Reply:
> Optimizing away I/Os is a firrtl issue, not really a Chisel issue imho

I think this is only true if FIRRTL should never optimize away I/Os but given it's an intentional feature, I think you need a Chisel-level user API to say either when you should, or shouldn't, optimize ports away. Making it such that ports defined as part of the "public API" seems like it could be convenient/intuitive but perhaps I'm wrong here.

Reply:
> I think this is only true if FIRRTL should never optimize away I/Os but given it's an intentional feature, I think you need a Chisel-level user API to say either when you should, or shouldn't, optimize ports away. Making it such that ports defined as part of the "public API" seems like it could be convenient/intuitive but perhaps I'm wrong here.

I guess my stance on this had always been that only toplevel I/Os need to be preserved.

However, I think the discussion of whether there should be a way to annotate I/Os as public API should be independent from the Instance API discussion. I just wanted to point out that during Chisel elaboration nothing is optimized away, so there is no _technical_ reason as far as I can see for Chisel users to be concerned about optimized I/Os in the context of the new Instance API.